Rover: add support for wing sails

IamPete1 opened the pull request:
This adds the long promised support for wing sails for rover sailboats. A new servo function (124) wing sail elevator has been added. All (i hope) the relevant places have been update to make sure it does the right thing. 

I have tested that the output does what I expected in SITL, I do not have a boat with a wing sail to test this with, We could add a wing sail model to the sailboat SITL but it would be a fair bit of work.

I think this control method will work to get us started, possibly once some more experience has been gained in the wild with wing sails it can be improved upon.

rmackay9 commented on libraries/SRV_Channel/SRV_Channel.h:
```diff
@@ -147,6 +147,7 @@ class SRV_Channel {
         k_LED_neopixel2         = 121,
         k_LED_neopixel3         = 122,
         k_LED_neopixel4         = 123,
+        K_wingsail_elevator     = 124,
```
little typo on the case of the "K".  should be a small "k"

IamPete1 replied:
fixed

rmackay9 commented on libraries/SRV_Channel/SRV_Channel.cpp:
```diff
@@ -65,7 +65,7 @@ const AP_Param::GroupInfo SRV_Channel::var_info[] = {
     // @DisplayName: Servo output function
     // @Description: Function assigned to this servo. Seeing this to Disabled(0) will setup this output for control by auto missions or MAVLink servo set commands. any other value will enable the corresponding function
-    // @Values: 0:Disabled,1:RCPassThru,2:Flap,3:Flap_auto,4:Aileron,6:mount_pan,7:mount_tilt,8:mount_roll,9:mount_open,10:camera_trigger,11:release,12:mount2_pan,13:mount2_tilt,14:mount2_roll,15:mount2_open,16:DifferentialSpoilerLeft1,17:DifferentialSpoilerRight1,19:Elevator,21:Rudder,24:FlaperonLeft,25:FlaperonRight,26:GroundSteering,27:Parachute,28:EPM,29:LandingGear,30:EngineRunEnable,31:Heli ...
+    // @Values: 0:Disabled,1:RCPassThru,2:Flap,3:Flap_auto,4:Aileron,6:mount_pan,7:mount_tilt,8:mount_roll,9:mount_open,10:camera_trigger,11:release,12:mount2_pan,13:mount2_tilt,14:mount2_roll,15:mount2_open,16:DifferentialSpoilerLeft1,17:DifferentialSpoilerRight1,19:Elevator,21:Rudder,24:FlaperonLeft,25:FlaperonRight,26:GroundSteering,27:Parachute,28:EPM,29:LandingGear,30:EngineRunEnable,31:Heli ...
```
let's make it "124:Wing Sail Elevator" or "124:WingSailElevator" to be more consistent with the case of other entries

IamPete1 replied:
fixed

rmackay9 commented on APMrover2/sailboat.cpp:
```diff
@@ -211,11 +224,26 @@ void Sailboat::get_throttle_and_mainsail_out(float desired_speed, float &throttl
+    //
+
+    // wing sails auto trim, we only need to reduce power if we are tipping over
+    wingsail_out = 100.0f - pid_offset;
```
Does this mean that the wing sail is nearly always fully left or right?  I'm probably misunderstanding something..

IamPete1 replied:
yeah, it is for use with a sail such as [this ](https://www.wired.com/images_blogs/autopia/2014/02/ff_saildrone_f.jpg) full elevator gives full power either tack, there is no need to trim it for wind direction.

IamPete1 replied:
it could be that for sailing upwind less elevator is more efficient, but this is probably to nuanced to implement until we do some proper testing

rmackay9 replied:
The elevator is really there to aim the sail right?  It's not the wagging of the sail that provides the thrust of course?  If the elevator is for aiming the sail then I think the elevator output should reach an equilibrium and that would mostly not be full left or right always.

IamPete1 replied:
exactly, the elevator just points the sail. The idea is that the user would set the servo min and max to a good equilibrium point for sailing on either tack.

rmackay9 replied:
I also strongly suspect this will need to be changed after a bit of testing.  Surely it's going to stall or turn the sail too far so the wind hits the back of the wing or something like this.  Anyway, I won't hold up the PR for this 'cuz as you say, it's a start.

rmackay9 commented:
Really great addition!

IamPete1 commented:
rebased

IamPete1 commented:
Rebased and fixed a manual mode bug found by a user testing this. This also needs a fix for https://github.com/ArduPilot/ardupilot/issues/12744 before it can be used

IamPete1 commented:
@rmackay9 more stuff left over from GSoC, this is safe now we have more servo functions, has been tested by Karl although not extensively. As I said at the top: 'I think this control method will work to get us started, possibly once some more experience has been gained in the wild with wing sails it can be improved upon.'

rmackay9 commented on APMrover2/sailboat.cpp:
```diff
@@ -207,11 +208,23 @@ void Sailboat::get_throttle_and_mainsail_out(float desired_speed, float &throttl
     // mainsail control
     //
-    // if we are motoring or attempting to reverse relax the sail
-    if (motor_state == UseMotor::USE_MOTOR_ALWAYS || !is_positive(desired_speed)) {
```
Is the removal of the motor_state and then lower down the pid_offset intentional?  These two changes seem unrelated..

IamPete1 replied:
They have been moved up,  motor_state  is on line 212 and the PID offset has been moved so we can use it for both the mainsail and wingsail. The reason for this re-jig is that we can go backwards with a wing sail (no promising it will work well, but it is possible) but not a 'normal' sail.  Thanks for taking a look

rmackay9 commented:
As mentioned this look pretty OK, just a couple of changes that I think might be related to a bad rebase.

rmackay9 commented:
Merged, thanks!
